## User Behavioral Analytics on Videos Dataset

In [1]:
#Importing libraries
import pandas as pd
import matplotlib.pyplot as plt

In [2]:
#Reading Data from csv
dataset= pd.read_csv('cdnelastic_VoD.csv')

In [3]:
dataset

coordinates  timetoserv  uafamily      sid contenttype  statuscode  \
0                1.0    0.022135      16.0     3871   video/mp4         200   
1                1.0    0.000248      16.0     3323   video/mp4         200   
2                1.0    0.000221      16.0     3930   video/mp4         200   
3                1.0    0.001468      16.0     3945   video/mp4         200   
4                1.0    0.000265      16.0     3739   video/mp4         200   
...              ...         ...       ...      ...         ...         ...   
3696289          0.0    0.010339      12.0  1280459  video/mpeg         206   
3696290       4823.0    0.000293      12.0  1280455  video/mpeg         206   
3696291       1792.0    0.000358      16.0  1281366   video/mp4         200   
3696292          0.0    0.000317      12.0  1281441  video/mpeg         206   
3696293          0.0    0.031679      12.0  1281495   video/mp4         206   

         uid  protocol   hit                @timestamp  contentpackage  host  \
0        NaN  HTTP/1.1  miss  2088-05-13T09:20:46.000Z            55.0    15   
1        NaN  HTTP/1.1   hit  2088-05-13T09:20:55.000Z            55.0    15   
2        NaN  HTTP/1.1   hit  2088-05-13T09:23:34.000Z            59.0    15   
3        NaN  HTTP/1.1   hit  2088-05-13T09:23:34.000Z            59.0    15   
4        NaN  HTTP/1.1   hit  2088-05-13T09:23:36.000Z            59.0    15   
...      ...       ...   ...                       ...             ...   ...   
3696289  NaN  HTTP/1.1  miss  2088-05-19T05:49:45.000Z          3938.0     4   
3696290  NaN  HTTP/1.1   hit  2088-05-19T05:51:33.000Z          2829.0     4   
3696291  NaN  HTTP/1.1   hit  2088-05-19T06:52:35.000Z          3821.0    15   
3696292  NaN  HTTP/1.1   hit  2088-05-19T06:57:49.000Z          2753.0     4   
3696293  NaN  HTTP/1.1  miss  2088-05-19T06:58:18.000Z          2753.0     4   

         assetnumber  livechannel  contentlength  
0               54.0          NaN   33963.682540  
1               54.0          NaN     518.730159  
2               58.0          NaN     512.809524  
3               58.0          NaN    7758.412698  
4               58.0          NaN     518.666667  
...              ...          ...            ...  
3696289       3472.0          NaN    1543.206349  
3696290       2553.0          NaN    1579.206349  
3696291       2357.0          NaN    4768.746032  
3696292       2501.0          NaN    1541.666667  
3696293       2501.0          NaN    5918.857143  

[3696294 rows x 15 columns]

After the data was loaded, We look at what columns we have and their types.

In [4]:
dataset.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3696294 entries, 0 to 3696293
Data columns (total 15 columns):
 #   Column          Dtype  
---  ------          -----  
 0   coordinates     float64
 1   timetoserv      float64
 2   uafamily        float64
 3   sid             int64  
 4   contenttype     object 
 5   statuscode      int64  
 6   uid             float64
 7   protocol        object 
 8   hit             object 
 9   @timestamp      object 
 10  contentpackage  float64
 11  host            int64  
 12  assetnumber     float64
 13  livechannel     float64
 14  contentlength   float64
dtypes: float64(8), int64(3), object(4)
memory usage: 423.0+ MB


In [5]:
dataset.isna().sum()

coordinates           550
timetoserv              0
uafamily           731034
sid                     0
contenttype             0
statuscode              0
uid               3426337
protocol                0
hit                     0
@timestamp              0
contentpackage          0
host                    0
assetnumber             0
livechannel       3696294
contentlength           0
dtype: int64

My focus will be mainly on SID, uid, contentpackage, for uid, I have many NaN values, so I will drop them to have a better view of the data.

In [19]:
dataset = dataset.dropna(subset=['uid'])
dataset = dataset.dropna(subset=['coordinates'])

In [5]:
dataset.isna().sum()

coordinates           77
timetoserv             0
uafamily            2543
sid                    0
contenttype            0
statuscode             0
uid                    0
protocol               0
hit                    0
@timestamp             0
contentpackage         0
host                   0
assetnumber            0
livechannel       269957
contentlength          0
dtype: int64

I will be dropping the following columns as I see they will not be necessary for this project: timetoserv, uafamily, protocol, hit, host, livechannel.

In [17]:
columnsdrop = ['timetoserv', 'uafamily', 'protocol', 'hit', 'host', 'livechannel']
dataset = dataset.drop(columnsdrop, axis=1)

In [25]:
dataset = dataset.reset_index(drop=True)

In [39]:
#changing name of contentpackage to contentID
dataset.rename(columns = {'contentpackage' : 'contentid'}, inplace=True)

Now, let us look at the data and get an idea after some basic cleaning

In [40]:
dataset.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 269880 entries, 0 to 269879
Data columns (total 9 columns):
 #   Column         Non-Null Count   Dtype  
---  ------         --------------   -----  
 0   coordinates    269880 non-null  float64
 1   sid            269880 non-null  int64  
 2   contenttype    269880 non-null  object 
 3   statuscode     269880 non-null  int64  
 4   uid            269880 non-null  float64
 5   @timestamp     269880 non-null  object 
 6   contentid      269880 non-null  float64
 7   assetnumber    269880 non-null  float64
 8   contentlength  269880 non-null  float64
dtypes: float64(5), int64(2), object(2)
memory usage: 18.5+ MB


I will now count how many user Ids we have, how many session Ids, and how many content IDs.

In [31]:
#UserID count
dataset['uid'].value_counts()
uidCount = len(dataset['uid'].value_counts())
print(uidCount)

3668


In [33]:
#SessionID count
dataset['sid'].value_counts()
sidCount = len(dataset['sid'].value_counts())
print(sidCount)

15336


In [41]:
#content id count
dataset['contentid'].value_counts()
sidCount = len(dataset['contentid'].value_counts())
print(sidCount)

3422


In [42]:
dataset['contenttype'].value_counts()

video/mp4                          198187
video/mpeg                          51330
text/xml                            19241
application/dash+xml                 1011
application/ttaf+xml                  103
application/json; charset=utf-8         8
Name: contenttype, dtype: int64

So in this dataset, we have activity of 3668 users, 15336 sessions, and 3422 content to be viewed.